python/ray/air/util/data_batch_conversion.py

```diff
@@ -148,6 +148,7 @@ def _cast_ndarray_columns_to_tensor_extension(df: pd.DataFrame) -> pd.DataFrame:
     # TODO(Clark): Once Pandas supports registering extension types for type
     # inference on construction, implement as much for NumPy ndarrays and remove
     # this. See https://github.com/pandas-dev/pandas/issues/41848
+    df = df.copy()
```

Comment:
cc @clarkzinzow what does this mean for our memory usage?

Reply:
I don't think this is the right solution to those warnings, the `SettingWithCopyWarning` is issued when you mutate a copy of a dataframe and then thrown it away, instead of the original dataframe. This commonly happens when incorrectly mutating a column or row slice.

@peytondmurray Could you point me to the case in which this happens?

Reply:
And yeah we'd only want to resort to this blanket-copy if no other solution exists, since this can be super expensive, and even then, we would only want to copy the dataframe if this cast actually happens (i.e. if that if block is one or more times in the for loop).

Reply:
@clarkzinzow The `tfx_tabular_train_to_serve.ipynb` that is part of the AIR examples was producing these errors. I just tried reproducing them after rebasing onto `master` but I wasn't able to, I assume because something changed elsewhere such that views of dataframes are no longer being passed into these functions. Given that, maybe it's just best to remove these copies given that they can be very expensive. I'll remove these copies now.

Anyway, if you pass in a view of a dataframe into either `_cast_ndarray_columns_to_tensor_extension` or `_cast_tensor_columns_to_ndarrays` in `ray/air/util/data_batch_conversion.py` these casts won't propagate back to the original dataframe, so we should still be careful if we aren't going to copy here.

Reply:
I also noticed that `DataFrame` has an `_is_view` property which is in principle not public, but could be useful for identifying when a copy is necessary, for future reference.

Reply:
@peytondmurray Ah I get what's happening here: a batch row-slice view is being passed to this function, which is then mutating the column, which is causing an erroneous `SettingWithCopyWarning` to be emitted. In reality, we don't care if the original DataFrame is mutated since we're operating on distinct batches and the original DataFrame can be freely used as scratch space, and if only the batch slice is actually mutated, that's also fine (for correctness). In fact, for the sake of performance, mutating the existing DataFrame is preferred if it's possible.

Reply:
Great - I think we can just use the [`option_context`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.option_context.html) context manager here to temporarily suppress the warning.